distributed/worker_state_machine.py

```diff
@@ -1610,17 +1635,26 @@ def _select_keys_for_gather(
         """Helper of _ensure_communicating.
 
         Fetch all tasks that are replicated on the target worker within a single
-        message, up to transfer_message_target_bytes.
+        message, up to transfer_message_target_bytes or until we reach the limit
+        for the size of incoming data transfers.
         """
         to_gather: list[TaskState] = []
         total_nbytes = 0
 
         while available:
             ts = available.peek()
             # The top-priority task is fetched regardless of its size
-            if (
-                to_gather
-                and total_nbytes + ts.get_nbytes() > self.transfer_message_target_bytes
+            gather_at_least_one_task = self.transfer_incoming_bytes or to_gather
+            exceed_message_target = (
+                total_nbytes + ts.get_nbytes() > self.transfer_message_target_bytes
+            )
+            exceed_bytes_limit = (
+                self.transfer_incoming_bytes_limit is not None
+                and self.transfer_incoming_bytes + total_nbytes + ts.get_nbytes()
+                > self.transfer_incoming_bytes_limit
+            )
+            if gather_at_least_one_task and (
+                exceed_message_target or exceed_bytes_limit
```

Comment:
How about

```python
        if self.transfer_incoming_bytes_limit:
            bytes_left_to_fetch = min(
                self.transfer_incoming_bytes_limit - self.transfer_incoming_bytes,
                self.transfer_message_target_bytes,
            )
        else:
            bytes_left_to_fetch = self.transfer_message_target_bytes

        while available:
            ts = available.peek()
            if (
                # If there is no other traffic, the top priority task may be
                # fetched regardless of its size
                to_gather or self.transfer_incoming_bytes
            ) and total_nbytes + ts.get_nbytes() > bytes_left_to_fetch:
                break
            for worker in ts.who_has:
                # This also effectively pops from available
                self.data_needed[worker].remove(ts)
            to_gather.append(ts)
            total_nbytes += ts.get_nbytes()

        return to_gather, total_nbytes
```

At least subjectively this seems simpler.

Reply:
Agreed, this looks simpler.